Address Issue #171: Validate Generated Addresses Against Expected Ones in JS Client Library

landabaso opened the pull request:
This PR aims to address issue #171 by implementing an enhanced validation method in the JavaScript client library. It enables the comparison of the generated wallet addresses with the expected ones, leveraging "@bitcoinerlab/descriptors". Here are the key updates:

1. Updated package.json to incorporate new dependencies "@bitcoinerlab/descriptors" and "@bitcoinerlab/secp256k1", and the latest version of "bitcoinjs-lib" to "^6.1.3".

2. Introduced a new test in appClient.test.ts to validate the generated address or throw an error if it doesn't match the expected one.

3. Made modifications in appClient.ts to integrate new imports. Changes have been made to the `AppClient` class to use these libraries for address validation. Some previous validations were removed (see below for further discussion).

Note that this PR utilizes "@bitcoinerlab/descriptors" for validation purposes. It offers an effective approximation for validating addresses. You can view the [extensive set of tests that bitcoinerlab passes here](https://github.com/bitcoinerlab/descriptors/tree/main/test/fixtures).

However, it's worth noting that "@bitcoinerlab/descriptors" does not yet support Tapscript yet. As a result, this third-party validation may not be available for certain types of descriptors. The code takes this into account and handles these cases appropriately.

In `getWalletAddress`, the previous validation method has been replaced with the new approach comparing generated addresses to expected ones using "@bitcoinerlab/descriptors".

Meanwhile, I'd like to discuss the removal of validation in the `registerWallet` and `signPSBT` methods of the `AppClient` class. An alternative approach could be to generate addresses for dummy keys and compare them using the `validateAddress` method. However, this might be excessive, potentially slow down performance, and there might be other drawbacks I'm not considering. I look forward to your comments on this proposal.

Regarding the versioning in package.json, I've decided not to update the version number. It appears that the version number of the npm package has been synchronized with the version of the entire repository which includes Rust and Python clients as well. This leads to a situation where a fix in TypeScript would have to wait for a new release version of the whole project before it could be submitted to npm.

I look forward to your feedback.

Closes #171.

Commit:
Add setInputWitnessScript and getInputWitnessScript methods

landabaso committed:
Fix setInputWitnessScript

landabaso committed:
Create a top level package.json file so that this repo can be installed using npm install

landabaso committed:
Better run the build script in the child's directory

landabaso committed:
Merge branch 'LedgerHQ:develop' into develop

landabaso committed:
Change amount type from Buffer to number in setInputWitnessUtxo / getInputWitnessUtxo

landabaso committed:
Remove unnecessary tiny-secp256k1 dependency

landabaso committed:
Add support for importing BitcoinJS Psbt for Ledger signing

landabaso committed:
Remove tiny-secp256k1 dependency from package-lock.js

landabaso committed:
Return the 'this' object in fromBitcoinJS to allow chaining

landabaso committed:
Merge branch 'bitcoinjslib_compat' into develop

landabaso committed:
Merge from import-from-bitcoinjs

landabaso committed:
Expose DefaultDescriptorTemplate

landabaso committed:
Adapt package.json so that this can be published to npm (to be used internally) without interfering with the original package.json

landabaso committed:
Add proper credit to Ledger for this module

landabaso committed:
Merge remote-tracking branch 'ledgers-remote/develop' into develop

landabaso committed:
Merge remote-tracking branch 'ledgers-remote/develop' into develop

landabaso committed:
Sync wrt ledgers remote

landabaso committed:
Remove package-lock.json files (to be in synch with ledger-bitcoin package)

landabaso committed:
Merge branch 'LedgerHQ:develop' into develop

landabaso committed:
Address #171: Compare generated addresses with expected ones

bigspider commented on bitcoin_client_js/src/lib/appClient.ts:
```diff
@@ -402,12 +403,69 @@ export class AppClient {
+      const keyPath = walletPolicy.keys[i] + '/' + change + '/' + addressIndex;
+      expression = expression
+        .replace('@' + i + '/**', keyPath)
+        .replace('@' + i + '/<0;1>', keyPath);
```
This would not work for policies containing '<M;N>' different than '<0;1>', which are supported.
`/<M;N>' should be replaced with `/M` if `change == 0`. with `/N` otherwise.

I should probably have typed the `change` parameter as `boolean` − this is my mistake (probably because wallet policies weren't clearly defined yet when I wrote the initial clients). Will think how to fix it for a future release.

bigspider reviewed (changes requested):
Thanks for the PR, great to be able to validate addresses at least on SegWit!

Few comments to align with the other libraries:

- The approach I used in the other libraries is to check the address _both_ at the end of get_wallet_address (no performance impact), but also at the end of register_wallet. This could have a slight performance impact, as generating an address is explicit extra work. however, this is probably negligible in practice.
- `validateAddress` can completely replace the existing `validatePolicy`, since both the bugs that are checked are for SegWit miniscript. The `containsA` helper can also be deleted.

Feel free to bump the package version to `0.2.3`, there's no plan to keep the version aligned across clients.

landabaso commented:
> * The approach I used in the other libraries is to check the address _both_ at the end of get_wallet_address (no performance impact), but also at the end of *register_wallet*.

I just realised that adding the check mentioned above will break the test for `registerWallet`. See the [code for this test](https://github.com/LedgerHQ/app-bitcoin-new/blob/a1fe503105658b4b0e0e40e833a623f992f0385d/bitcoin_client_js/src/__tests__/appClient.test.ts#L263) here:

```javascript
  it("can register a miniscript wallet", async () => {
    const walletPolicy = new WalletPolicy(
      "Decaying 3-of-3",
      "wsh(thresh(3,pk(@0/**),s:pk(@1/**),s:pk(@2/**),sln:older(12960)))",
      [
        "[76223a6e/48'/1'/0'/2']tpubDE7NQymr4AFtewpAsWtnreyq9ghkzQBXpCZjWLFVRAvnbf7vya2eMTvT2fPapNqL8SuVvLQdbUbMfWLVDCZKnsEBqp6UK93QEzL8Ck23AwF",
        "[f5acc2fd/48'/1'/0'/2']tpubDFAqEGNyad35aBCKUAXbQGDjdVhNueno5ZZVEn3sQbW5ci457gLR7HyTmHBg93oourBssgUxuWz1jX5uhc1qaqFo9VsybY1J5FuedLfm4dK",
      ]
    );

    const automation = JSON.parse(fs.readFileSync('src/__tests__/automations/register_wallet_accept.json').toString());
    await setSpeculosAutomation(transport, automation);

    const [walletId, walletHmac] = await app.registerWallet(walletPolicy);

    expect(walletId).toEqual(walletPolicy.getId());
    expect(walletHmac.length).toEqual(32);
  }, 60000);
```

It's worth mentioning that the template requires 3 keys, but only 2 keys are provided. Generating an address with an incomplete set of keys isn't feasible. Is this a legitimate policy anyway? If so, how should we approach this circumstance?

landabaso committed:
Addressed review comments: validate addresses at getWalletAddress and end of registerWallet; validateAddress replaces validatePolicy; bumped package version; correctly deal with <M;N> format (added a test using this format); Fixed incomplete keys in registerWallet test

landabaso commented:
Hi @bigspider , I've made the changes you suggested. Specifically, I've:

* Bumped the package version to 0.2.3.
* Added address validation at the end of `registerWallet`.
* Completely replaced `validatePolicy` with `validateAddress` and removed `containsA`.
* Adjusted the code for correctly parsing the `<M;N>` format - I had initially misunderstood its interpretation. Included a test case using this format for better clarity.
* Added a random third key to the registerWallet test to avoid the incomplete keys issue explained in my previous comment. Please, confirm this is ok.

Please let me know if you have any additional feedback.

landabaso committed:
Fix issue with key replacement in address derivation

Revised the key replacement logic in the `validateAddress` function to prevent misinterpretation of key indices. The loop now iterates in reverse order to avoid scenarios where, for example, @10 is mistakenly replaced as @1, leaving an extra 0. This change ensures that the correct keys are replaced when deriving the wallet address.

bigspider reviewed (dismissed):
Well spotted for the incorrect policy!
Fixed in #188, including the defective test.

Only noted a small nit.

Happy to merge it once rebased on the current `develop` branch, thanks a lot!

landabaso committed:
Fix typo and rebase onto develop branch

landabaso commented on bitcoin_client_js/src/lib/appClient.ts:
```diff
@@ -402,34 +407,75 @@ export class AppClient {
-        // the `a:` fragment.
-        throw new Error("Please update your Ledger Bitcoin app.")
-      }
+  /* Performs any additional check on the genetated address before returning it.*/
```
Oups, it looks like the rebase+commit did not go as expected. Hang on while I figure out how to solve this.

landabaso committed:
Merge branch 'develop' into improved-validation

landabaso commented:
This PR is being closed because it has been replaced by PR #189 due to issues with a faulty rebase operation. Please refer to PR #189 for the latest changes.
